clients/src/main/java/org/apache/kafka/clients/admin/internals/PartitionLeaderStrategy.java

```diff
@@ -0,0 +1,190 @@
+                    "Failed to fetch metadata for partition " + tp + " due to invalid topic `" + topic + "`",
+                    Collections.singleton(topic)));
+                break;
+
+            default:
+                log.error("Received unexpected error for topic {} in `Metadata` response", topic,
+                    topicError.exception());
+                failAllPartitionsForTopic(topic, requestPartitions, failed, tp -> topicError.exception(
+                    "Failed to fetch metadata for partition " + tp + " due to unexpected error for topic `" + topic + "`"));
+        }
+    }
+
+    private void failAllPartitionsForTopic(
+        String topic,
+        Set<TopicPartition> partitions,
+        Map<TopicPartition, Throwable> failed,
+        Function<TopicPartition, Throwable> exceptionGenerator
+    ) {
+        partitions.stream().filter(tp -> tp.topic().equals(topic)).forEach(tp -> {
+            failed.put(tp, exceptionGenerator.apply(tp));
+        });
+    }
+
+    private void handlePartitionError(
+        TopicPartition topicPartition,
+        Errors partitionError,
+        Map<TopicPartition, Throwable> failed
+    ) {
+        switch (partitionError) {
+            case NOT_LEADER_OR_FOLLOWER:
+            case REPLICA_NOT_AVAILABLE:
+            case LEADER_NOT_AVAILABLE:
+            case BROKER_NOT_AVAILABLE:
+            case KAFKA_STORAGE_ERROR:
+                log.debug("Metadata request for partition {} returned partition-level error {}. Will retry",
+                    topicPartition, partitionError);
+                break;
+
+            default:
+                log.error("Received unexpected error for partition {} in `Metadata` response",
+                    topicPartition, partitionError.exception());
+                failed.put(topicPartition, partitionError.exception(
+                    "Unexpected error during metadata lookup for " + topicPartition));
+        }
+    }
+
+    @Override
+    public LookupResult<TopicPartition> handleResponse(
+        Set<TopicPartition> requestPartitions,
+        AbstractResponse abstractResponse
+    ) {
+        MetadataResponse response = (MetadataResponse) abstractResponse;
+        Map<TopicPartition, Throwable> failed = new HashMap<>();
+        Map<TopicPartition, Integer> mapped = new HashMap<>();
+
+        for (MetadataResponse.TopicMetadata topicMetadata : response.topicMetadata()) {
+            Errors topicError = topicMetadata.error();
+            if (topicError != Errors.NONE) {
+                handleTopicError(topicMetadata.topic(), topicError, requestPartitions, failed);
+                continue;
+            }
+
+            for (MetadataResponse.PartitionMetadata partitionMetadata : topicMetadata.partitionMetadata()) {
+                TopicPartition topicPartition = partitionMetadata.topicPartition;
+                Errors partitionError = partitionMetadata.error;
+
+
+                if (!topicPartitions.contains(topicPartition)) {
```

Comment:
`topicPartitions` should be equal to `requestPartitions` but it seems to me using `requestPartitions` is more acceptable for this check since it can be consistent with line#158 (handleTopicError(topicMetadata.topic(), topicError, ```requestPartitions```, failed);)

Reply:
Good point. I agree.